ui/lib/apps/TopSQL/pages/List/List.tsx

```diff
@@ -51,92 +58,90 @@ export function TopSQLList() {
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [containerWidth, timeRange])
-
-  const handleUpdateTopSQLData = useCallback(() => {
-    // Update container width when refresh, then trigger the update of time window size to refresh.
-    const cw = containerRef.current?.offsetWidth || 0
-    if (cw !== containerWidth) {
-      setContainerWidth(containerRef.current?.offsetWidth || 0)
-      return
-    }
+  const {
+    instances,
+    isLoading: isInstancesLoading,
+    updateInstances,
+  } = useInstances()
 
-    const timestamps = calcTimeRange(timeRange)
-    setQueryTimestampRange(timestamps)
+  const updateData = useCallback(
+    (_instance: TopsqlInstanceItem | null, _timeRange: TimeRange) => {
+      if (!_instance) {
+        return
+      }
 
-    if (!instance || !timeWindowSize) {
-      return
-    }
+      // Update container width when refresh, then trigger the update of time window size to refresh.
+      const cw = containerRef.current?.offsetWidth || containerWidth
+      if (cw !== containerWidth) {
+        setContainerWidth(containerRef.current?.offsetWidth || 0)
+      }
 
-    updateTopSQLData({ instance, timestamps, timeWindowSize, topN })
-  }, [
-    instance,
-    timeRange,
-    timeWindowSize,
-    containerRef,
-    containerWidth,
-    updateTopSQLData,
-  ])
+      const timestamps = toTimeRangeValue(_timeRange)
+      const _timeWindowSize = computeTimeWindowSize(cw, timestamps)
 
-  // fetch data when instance id / time window size update
-  useEffect(() => {
-    handleUpdateTopSQLData()
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [instance, timeWindowSize])
+      updateInstances(_timeRange)
+      updateTopSQLData({
+        instance: _instance,
+        timestamps,
+        timeWindowSize: _timeWindowSize,
+        topN,
+      })
+    },
+    [
+      containerRef,
+      containerWidth,
+      updateTopSQLData,
+      computeTimeWindowSize,
+      updateInstances,
+    ]
+  )
 
   const handleBrushEnd: BrushEndListener = useCallback(
     (v: BrushEvent) => {
       if (!v.x) {
         return
       }
 
+      let value: [number, number]
       const tr = v.x.map((d) => d / 1000)
       const delta = tr[1] - tr[0]
       if (delta < 60) {
         const offset = Math.floor(delta / 2)
-        const value: [number, number] = [
+        value = [
           Math.ceil(tr[0] + offset - 30),
           Math.floor(tr[1] - offset + 30),
         ]
-
-        setTimeRange({
-          type: 'absolute',
-          value,
-        })
-        telemetry.dndZoomIn(value)
-        return
+      } else {
+        value = [Math.ceil(tr[0]), Math.floor(tr[1])]
       }
 
-      const value: [number, number] = [Math.ceil(tr[0]), Math.floor(tr[1])]
-      setTimeRange({
+      const _timeRange: TimeRange = {
         type: 'absolute',
         value,
-      })
+      }
+
+      setTimeRange(_timeRange)
+      updateData(instance, _timeRange)
```

Comment:
Maybe you can let `updateData` to be an effect of:

- selected time range is changed
- selected instance is changed

This is how useEffect build for.

Refresh can be effectively implemented as the following:

```js
setTimeRange(r => [...r])
```

Reply:
Impressive! I was extremes of observer and event emit. Poor react again.